python3/bin/perfmon

```diff
@@ -0,0 +1,1620 @@
+            log_err(
+                "caught connection refused error: (%s) - restarting XAPI session"
+                % str(e)
+            )
+            restart_session = True
+        except urllib.error.HTTPError as e:
+            if e.code in (401, 500):
+                # Error getting rrd_updates: 401=Unauthorised, 500=Internal
+                # start new session
+                log_err("caught http.error: (%s) - restarting XAPI session" % str(e))
+                restart_session = True
+            else:
+                # Don't know why we got this error - crash, die and look at logs later
+                raise
+        except OSError as e:
+            # This happens if we send messages or
+            # read other-config:perfmon after xapi is restarted
+            log_err("caught connection error: (%s) - restarting XAPI session" % str(e))
+            restart_session = True
+
+        runs += 1
+        if maxruns is not None and runs >= maxruns:
+            break
+
+        # Force collection of cyclically referenced objects cos we don't
+        # trust GC to do it on its own
+        gc.collect()
+
+        # Sleep for interval + dither, exiting early if we recv a cmd
+        timeout = rand(interval, interval + dither)
+        cmdsock.settimeout(timeout)
+        try:
+            cmd = cmdsock.recv(cmdmaxlen)
+        except socket.timeout:
+            pass
+        else:
+            if cmd == "refresh":
+                # This forces a re-read of all the configs on the next loop
+                next_config_update = time.time()
+            elif cmd == "debug_mem":
+                debug_mem()
+            else:
+                log_err("received unhandled command %s" % cmd)
+
+        # continue to next run
+
+    return 0
+
+
+def sigterm_handler(sig, stack_frame): # pragma: no cover
+    _ = stack_frame  # unused: not sure if it'll be used later, passing pylint
+    log_err("Caught signal %d - exiting" % sig)
+    sys.exit(1)
+
+
+pidfile = "/var/run/perfmon.pid"
+
+if __name__ == "__main__": # pragma: no cover
+
+    # setup signal handler to print out notice when killed
+    signal.signal(signal.SIGTERM, sigterm_handler)
+
+    if "--daemon" in sys.argv[1:]:
+        sys.argv.remove("--daemon")
+        if os.fork() != 0:
+            sys.exit(0)
+        os.setsid()
+        # For /dev/null, `encoding` and `with` is not needed
+        # pylint: disable=unspecified-encoding, consider-using-with
+        sys.stdout = open("/dev/null", "w")
+        sys.stdin = open("/dev/null", "r")
+        sys.stderr = sys.stdout
+
+    # Exit if perfmon already running
+    if os.path.exists(pidfile):
+        with open(pidfile, encoding="utf-8") as file:
+            pid = file.read()
+
+        if os.path.exists("/proc/%s" % pid):
+            log_err("perfmon already running - exiting")
+            sys.exit(3)
+
+    try:
+        # Write out pidfile
+        with open(pidfile, "w", encoding="utf-8") as fd:
```

Comment:
This kind of read-and-write pattern seems not correct.
1. read from a file
2. check
3. write to the file

Reply:
Need to add a lock for the file operations.
As suggested, raise another ticket for this kind of refine work.
Please check CP-48628